test/filewriter.spec.ts

```diff
@@ -1,41 +1,88 @@
+
 describe("when ending SlpFileWriter", () => {
   it("should write data length to file", async () => {
-    const testFilePath = "slp/finalizedFrame.slp";
-
-    const slpFileWriter = new SlpFileWriter();
-    const slpFile = openSlpFile({ source: SlpInputSource.FILE, filePath: testFilePath });
-    const dataLength = slpFile.rawDataLength;
-    const dataPos = slpFile.rawDataPosition;
+    const { dataLength, newFilename } = runSlpFileWriter("slp/finalizedFrame.slp");
+    await new Promise((resolve) => {
+      setTimeout(() => {
+        const writtenDataLength = openSlpFile({ source: SlpInputSource.FILE, filePath: newFilename }).rawDataLength;
+        fs.unlinkSync(newFilename);
 
-    const testFd = fs.openSync(testFilePath, "r");
-    const newPos = pipeMessageSizes(testFd, dataPos, slpFileWriter);
+        expect(writtenDataLength).toBe(dataLength);
 
-    const newFilename = slpFileWriter.getCurrentFilename();
-    const buffer = Buffer.alloc(4);
+        resolve();
+      }, TIMEOUT_MS);
+    });
+  });
 
-    pipeAllEvents(testFd, newPos, dataPos + dataLength, slpFileWriter, slpFile.messageSizes);
+  it("should succeed if no display names or connect codes are available in game start", async () => {
+    const { newFilename } = runSlpFileWriter("slp/finalizedFrame.slp");
     await new Promise((resolve) => {
-      // On my machine, >100 is required to give the slpFile.ts "finish" callback time to execute.
-      // I thought a 'yield' 0 ms setTimout would allow the callback to execute, but that's not the case.
-      const timeoutMs = 1000;
+      setTimeout(() => {
+        const players = new SlippiGame(newFilename).getMetadata().players;
+        Object.keys(players).forEach((key) => {
+          const player = players[key];
+          players[key] = player.names;
+        });
+        fs.unlinkSync(newFilename);
+
+        expect(players).toEqual({ 0: { netplay: "", code: "" }, 1: { netplay: "", code: "" } });
 
+        resolve();
+      }, TIMEOUT_MS);
+    });
+  });
+
+  it("should write display name and connect codes to metadata if available in game start", async () => {
+    const { newFilename } = runSlpFileWriter("slp/displayNameAndConnectCodeInGameStart.slp");
+    await new Promise((resolve) => {
       setTimeout(() => {
-        const writtenDataLength = openSlpFile({ source: SlpInputSource.FILE, filePath: newFilename }).rawDataLength;
+        const players = new SlippiGame(newFilename).getMetadata().players;
+        Object.keys(players).forEach((key) => {
+          const player = players[key];
+          players[key] = player.names;
+        });
         fs.unlinkSync(newFilename);
 
-        expect(writtenDataLength).toBe(dataLength);
+        expect(players).toEqual({
+          0: { netplay: "ekans", code: "EKNS#442" },
+          1: { netplay: "gaR's uncle", code: "BAP#666" },
+          2: { netplay: "jmlee337", code: "JMLE#166" },
+          3: { netplay: "Mr.SuiSui", code: "SUI#244" },
```

Comment:
Again please make these `netplay.name` and `netplay.code`.

Reply:
these are already existing fields in MetadataType. Changing MetadataType fields would break compatibility with existing replay files.

In this light, do you still want me to make the corresponding change in slpFile.ts? That type is internal to the file and can be changed.

Reply:
Sadge